src/tools/rustfmt/src/parse/parser.rs

```diff
@@ -163,13 +163,21 @@ impl<'a> Parser<'a> {
     fn parse_crate_mod(&mut self) -> Result<ast::Crate, ParserError> {
         let mut parser = AssertUnwindSafe(&mut self.parser);
 
-        match catch_unwind(move || parser.parse_crate_mod()) {
-            Ok(Ok(k)) => Ok(k),
-            Ok(Err(db)) => {
+        // rustfmt doesn't use `run_compiler` like other tools, so it must emit
+        // any stashed diagnostics itself, otherwise the `DiagCtxt` will assert
+        // when dropped. The final result here combines the parsing result and
+        // the `emit_stashed_diagnostics` result.
+        let parse_res = catch_unwind(move || parser.parse_crate_mod());
+        let stashed_res = self.parser.dcx().emit_stashed_diagnostics();
```

Comment:
I'm wondering will this emit anything when using the silent emitter? I'm guessing it won't, but just want to double check.

Reply:
It shouldn't. `DiagCtxt::emit_stashed_diagnostics` calls `DiagCtxtInner:emit_diagnostic` on each diagnostic, which calls `emit_diagnostic` on the emitter held by the `DiagCtxt`. So if that's the silent emitter, then nothing will be printed.

Reply:
Thanks for the explanation 😁